server/src/main/java/org/elasticsearch/cluster/service/TransportVersionsFixupListener.java

```diff
@@ -0,0 +1,193 @@
+ * due to the master node not understanding cluster state with transport versions added in 8.8.0.
+ * Any nodes with the inferred placeholder cluster state is then refreshed with their actual transport version
+ */
+public class TransportVersionsFixupListener implements ClusterStateListener {
+
+    private static final Logger Log = LogManager.getLogger(TransportVersionsFixupListener.class);
+
+    private static final TimeValue RETRY_TIME = TimeValue.timeValueSeconds(30);
+    private static final int MAX_RETRIES = 120; // try for 1 hour
+
+    private final MasterServiceTaskQueue<NodeTransportVersionTask> taskQueue;
+    private final ClusterAdminClient client;
+    private final Scheduler scheduler;
+    private final Set<String> pendingNodes = Collections.synchronizedSet(new HashSet<>());
+
+    public TransportVersionsFixupListener(ClusterService service, ClusterAdminClient client, Scheduler scheduler) {
+        this(service.createTaskQueue("fixup-transport-versions", Priority.LOW, new TransportVersionUpdater()), client, scheduler);
+    }
+
+    TransportVersionsFixupListener(
+        MasterServiceTaskQueue<NodeTransportVersionTask> taskQueue,
+        ClusterAdminClient client,
+        Scheduler scheduler
+    ) {
+        this.taskQueue = taskQueue;
+        this.client = client;
+        this.scheduler = scheduler;
+    }
+
+    class NodeTransportVersionTask implements ClusterStateTaskListener {
+        private final Map<String, TransportVersion> results;
+        private final int retryNum;
+
+        NodeTransportVersionTask(Map<String, TransportVersion> results, int retryNum) {
+            this.results = results;
+            this.retryNum = retryNum;
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            Log.error("Could not apply transport version for nodes {} to cluster state", results.keySet(), e);
+            scheduleRetry(results.keySet(), retryNum);
+        }
+
+        public Map<String, TransportVersion> results() {
+            return results;
+        }
+    }
+
+    private static class TransportVersionUpdater implements ClusterStateTaskExecutor<NodeTransportVersionTask> {
+        @Override
+        public ClusterState execute(BatchExecutionContext<NodeTransportVersionTask> context) throws Exception {
+            ClusterState.Builder builder = ClusterState.builder(context.initialState());
+            for (var c : context.taskContexts()) {
+                for (var e : c.getTask().results().entrySet()) {
+                    // this node's transport version might have been updated already/node has gone away
+                    if (Objects.equals(builder.transportVersions().get(e.getKey()), INFERRED_TRANSPORT_VERSION)) {
+                        builder.putTransportVersion(e.getKey(), e.getValue());
+                    }
+                }
+            }
+            return builder.build();
```

Comment:
Can we return `context.initialState()` if there were no changes? Returning a new `ClusterState` instance always triggers a publication, even if it's identical to the previous one.